packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,494 @@
+   * @param trustedIssuers Array of n issuers whose identifier mappings will be used
+   * @return countsPerIssuer Array of number of identifiers returned per issuer
+   * @return identifiers Array (length == sum([0])) of identifiers
+   * @dev Adds identifier info to the arrays in order of provided trustedIssuers
+   * @dev Expectation that only one attestation exists per (identifier, issuer, account)
+   */
+  function lookupIdentifiers(address account, address[] calldata trustedIssuers)
+    external
+    view
+    returns (uint256[] memory countsPerIssuer, bytes32[] memory identifiers)
+  {
+    uint256 totalIdentifiers;
+    (totalIdentifiers, countsPerIssuer) = getNumIdentifiers(account, trustedIssuers);
+
+    identifiers = new bytes32[](totalIdentifiers);
+    bytes32[] memory identifiersPerIssuer;
+
+    uint256 currIndex = 0;
+
+    for (uint256 i = 0; i < trustedIssuers.length; i = i.add(1)) {
+      identifiersPerIssuer = addressToIdentifiers[account][trustedIssuers[i]];
+      for (uint256 j = 0; j < identifiersPerIssuer.length; j = j.add(1)) {
+        identifiers[currIndex] = identifiersPerIssuer[j];
+        currIndex = currIndex.add(1);
+      }
+    }
+    return (countsPerIssuer, identifiers);
+  }
+
+  /**
+   * @notice Validates the given attestation and signature
+   * @param identifier Hash of the identifier to be attested
+   * @param issuer Address of the attestation issuer
+   * @param account Address of the account being mapped to the identifier
+   * @param issuedOn Time at which the issuer issued the attestation in Unix time 
+   * @param signer Address of the signer of the attestation
+   * @param v The recovery id of the incoming ECDSA signature
+   * @param r Output value r of the ECDSA signature
+   * @param s Output value s of the ECDSA signature
+   * @dev Throws if attestation has been revoked
+   * @dev Throws if signer is not an authorized AttestationSigner of the issuer
+   */
+  function validateAttestationSig(
+    bytes32 identifier,
+    address issuer,
+    address account,
+    address signer,
+    uint64 issuedOn,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) public view {
+    require(
+      getAccounts().attestationSignerToAccount(signer) == issuer,
+      "Signer is not a currently authorized AttestationSigner for the issuer"
+    );
+    bytes32 structHash = keccak256(
+      abi.encode(EIP712_OWNERSHIP_ATTESTATION_TYPEHASH, identifier, issuer, account, issuedOn)
+    );
+    address guessedSigner = Signatures.getSignerOfTypedDataHash(
+      eip712DomainSeparator,
+      structHash,
+      v,
+      r,
+      s
+    );
+    require(guessedSigner == signer, "Signature is invalid");
+  }
+
+  function getUniqueAttestationHash(
+    bytes32 identifier,
+    address issuer,
+    address account,
+    address signer,
+    uint64 issuedOn
+  ) public pure returns (bytes32) {
+    return keccak256(abi.encode(identifier, issuer, account, signer, issuedOn));
```

Comment:
Highly recommend we use the EIP-712 hash here instead, as is done for [signature verification](https://github.com/celo-org/celo-monorepo/pull/9560/files#diff-6ce748d1111a6128ceb1e33a867d4628f747ccee8d741cd54aa2c63aaa701986R294-R296

Reply:
Bonus is that you could remove one hashing step from registration

Reply:
good idea, will modify!

Reply:
Hey one thing on this note, since there's a tiny difference in the two (the EIP-712 hash does not include the `signer` as is, whereas the revocation hash does). We can either add the `signer` field to the EIP712 attestation struct or remove it from the revocation hashes (or keep this as is with a slight difference). Regarding removing the signer, maybe there's some weird malicious behavior that’s possible where a signer targets one (account, identifier) and issues and revokes a bunch of attestations with various `issuedOn` times to be annoying (i.e. so it’s hard for the issuer/user to recreate and upload an attestation for that particular (account, identifier). I think I'm now leaning towards leaving the `signer` in the revocation hash, but undecided on whether or not we should include this in the EIP712 attestation data hash. Wdyt @nategraf @bowd ?